stingray/lightcurve.py

```diff
@@ -151,74 +183,146 @@ def __init__(self, time, counts, err=None, input_counts=True,
+            self._counts = np.asarray(counts)
+            self._counts_err = err
         else:
-            self.countrate = np.asarray(counts)[good]
-            self.counts = self.countrate * self.dt
-            self.counts_err = np.asarray(err)[good] * self.dt
-            self.countrate_err = np.asarray(err)[good]
-
-        self.meanrate = np.mean(self.countrate)
-        self.meancounts = np.mean(self.counts)
-        self.n = self.counts.shape[0]
-
+            self._countrate = np.asarray(counts)
+            self._countrate_err = err
+
+        if not skip_checks:
+            self.check_lightcurve()
+
+    @property
+    def mask(self):
+        if self._mask is None:
+            self._mask = create_gti_mask(self.time, self.gti, dt=self.dt)
+        return self._mask
+
+    @property
+    def n(self):
+        if self._n is None:
+            self._n = self.counts.shape[0]
+        return self._n
+
+    @property
+    def meanrate(self):
+        if self._meanrate is None:
+            self._meanrate = np.mean(self.countrate[self.mask])
+        return self._meanrate
+
+    @property
+    def meancounts(self):
+        if self._meancounts is None:
+            self._meancounts = np.mean(self.counts[self.mask])
+        return self._meancounts
+
+    @property
+    def counts(self):
+        if self._counts is None:
+            self._counts = self.countrate  * self.dt
+        return self._counts
+
+    @counts.setter
+    def counts(self, value):
+        self._counts = value
+
+    @property
+    def counts_err(self):
+        if self._counts_err is None and self._countrate_err is not None:
+            self._counts_err = self._countrate_err * self.dt
+        elif self._counts_err is None:
+            self._counts_err = poisson_symmetrical_errors(self.counts)
+        return self._counts_err
+
+    @counts_err.setter
+    def counts_err(self, value):
+        self._counts_err = value
+
+    @property
+    def countrate(self):
+        if self._countrate is None:
+            self._countrate = self.counts  / self.dt
+        return self._countrate
+
+    @countrate.setter
+    def countrate(self, value):
+        self._countrate = value
+
+    @property
+    def countrate_err(self):
+        if self._countrate_err is None:
+            self._countrate_err = self.counts_err  / self.dt
+        return self._countrate_err
+
+    @countrate_err.setter
+    def countrate_err(self, value):
+        self._countrate_err = value
+
+    @property
+    def bin_lo(self):
+        if self._bin_lo is None:
+            self._bin_lo = self.time - 0.5 * self.dt
+        return self._bin_lo
+
+    @property
+    def bin_hi(self):
+        if self._bin_hi is None:
+            self._bin_hi = self.time + 0.5 * self.dt
+        return self._bin_hi
+
+    def check_lightcurve(self):
+        """Make various checks on the lightcurve.
+
+        It can be slow, use it if you are not sure about your
+        input data.
+        """
```

Comment:
Could some of the tests implemented in `__init__` with `skip_checks` be collated here?

Reply:
There are checks that are good to run at the end of Lightcurve creation (those in this function), and others that need the initial values of time and counts, before detecting possible problems (run with skip_checks). This solution is not super clear, I admit, but it seemed like a good compromise

Reply:
Fair enough. It makes me wonder, though, if it's worth shoving all the earlier checks into its own method, too, just for readability of the `__init__` method?

Reply:
Done.